tc/autotuner/options_cache-inl.h

```diff
@@ -0,0 +1,316 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdint>
+#include <fstream>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <llvm/ADT/Optional.h>
+
+#include <version.h>
+
+#include "tc/core/tensor.h"
+#include "tc/core/utils/math.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace autotune {
+
+bool OptionsCacheKey::operator==(const OptionsCacheKey& other) const {
+  if (id != other.id) {
+    return false;
+  }
+  if (deviceStr != other.deviceStr) {
+    return false;
+  }
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    if (not(inputs[i] == other.inputs[i])) {
+      return false;
+    }
+  }
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    if (not(outputs[i] == other.outputs[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+OptionsCacheKeyProto OptionsCacheKey::toProtobuf() const {
+  OptionsCacheKeyProto buf_key;
+  buf_key.set_id(id);
+  for (const auto& in : inputs) {
+    auto pin = buf_key.add_inputs();
+    *pin = in.toProtobuf();
+  }
+  for (const auto& out : outputs) {
+    auto pout = buf_key.add_outputs();
+    *pout = out.toProtobuf();
+  }
+  buf_key.set_device_str(deviceStr);
+  buf_key.set_git_version("");
```

Comment:
git_version seems to be never modified by anything else in this PR, it will likely remain empty, and it isn't a likely intended behavior